### PR TITLE
feat: InputNotSorted lint rule

### PR DESCRIPTION
**Pull request by adthrasher**
This pull request adds a new rule to `wdl-lint`.

- **Rule Name**: `InputNotSorted`

Implements a rule to ensure that input declarations are sorted based on our opinionated sort.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

Rule specific checks:

- [x] You have added the rule as an entry within `RULES.md`.
- [x] You have added the rule to the `rules()` function in `wdl-lint/src/lib.rs`.
- [x] You have added a test cases in `wdl-lint/tests/lints` that covers every 
      possible diagnostic emitted for the rule within the file where the rule  
      is implemented.
- [x] You have run `wdl-gauntlet --refresh` to ensure that there are no 
      unintended changes to the baseline configuration file (`Gauntlet.toml`).
- [x] You have run `wdl-gauntlet --refresh --arena` to ensure that all of the 
      rules added/removed are now reflected in the baseline configuration file 
      (`Arena.toml`).

### CODE REVIEW

File: wdl-ast/src/v1/decls.rs [hunk 1]
```diff
@@ -545,6 +656,46 @@ impl fmt::Display for Type {
...
+}
+
+impl Ord for Type {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        compare_types(self, other)
+    }
+}
+
+/// Compare all variants of Type.
+fn compare_types(a: &Type, b: &Type) -> std::cmp::Ordering {
```
Comment (a-frantz) [inline wdl-ast/src/v1/decls.rs:672]: This is really hard to follow. Can it be rewritten using Rust `match` syntax?

File: wdl-lint/src/rules/input_not_sorted.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    fn explanation(&self) -> &'static str {
+        "Each input declaration section should be sorted. Enforces an opinionated sorting. First \
+         sorts by 1. required inputs, 2. optional inputs without defaults, 3. optional inputs with \
+         defaults, and 4. non-optional inputs with a default value. Then by the type: 1. File, 2. \
+         Array[*]+, 3. Array[*], 4. struct, 5. Object, 6. Map[*, *], 7. Pair[*, *], 8. String, 9. \
+         Boolean, 10. Float, 11. Int. For ordering of the same compound type (Array[*], Map[*, *], \
+         Pair[*, *]), drop the outermost type (Array, Map, etc.) and recursively apply above \
+         sorting on the first inner type *, with ties broken by the second inner type. Continue \
+         this pattern as far as possible. Once this ordering is satisfied, it is up to the \
+         developer for final order of inputs of the same type."
```
Comment (a-frantz) [inline wdl-lint/src/rules/input_not_sorted.rs:49]:
```suggestion
        "Each input declaration section should be sorted. This rule enforces an opinionated sorting. First \
         sorts by 1. required inputs, 2. optional inputs without defaults, 3. optional inputs with \
         defaults, and 4. inputs with a default value. Then by the type: 1. File, 2. \
         Array[*]+, 3. Array[*], 4. struct, 5. Object, 6. Map[*, *], 7. Pair[*, *], 8. String, 9. \
         Boolean, 10. Float, 11. Int. For ordering of the same compound type (Array[*], Map[*, *], \
         Pair[*, *]), drop the outermost type (Array, Map, etc.) and recursively apply above \
         sorting on the first inner type *, with ties broken by the second inner type. Continue \
         this pattern as far as possible. Once this ordering is satisfied, it is up to the \
         developer for final order of inputs of the same type."
```

File: wdl-lint/src/rules/input_not_sorted.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style])
```
Comment (a-frantz) [inline wdl-lint/src/rules/input_not_sorted.rs:53]: Should we add `Clarity`?

File: wdl-ast/src/v1/decls.rs (hunk 1 above)
Comment (adthrasher) [reply wdl-ast/src/v1/decls.rs:672]: Excellent suggestion! I learned something new about `match` and it improves the readability here.

File: wdl-lint/src/rules/input_not_sorted.rs (hunk 3 above)
Comment (adthrasher) [reply wdl-lint/src/rules/input_not_sorted.rs:53]: Probably. The only thing listed in #11 is `sorting`, which we haven't utilized.

Comment (a-frantz) [reply wdl-lint/src/rules/input_not_sorted.rs]: IMO `Sorting` would be a good addition here and also to `ImportSort`. I imagine the same people who don't intend to follow `ImportSort` won't follow `InputSort` either, so might as well give them a shorthand?

**Review by a-frantz — APPROVED** [review]
